### ISSUE TITLE
reproduce the result

### ISSUE THREAD

**Issue by zxduan-r** [skyhehe123/SA-SSD#5]
I run the script according to the readme.md. But I use the latest version of spconv. I got the result as follows:
AP@0.70, 0.70, 0.70:
bbox AP:86.75, 67.33, 60.57
bev  AP:89.95, 70.55, 65.85
3d   AP:74.24, 54.92, 50.28
aos  AP:86.56, 66.95, 60.12
Car AP@0.70, 0.50, 0.50:
bbox AP:86.75, 67.33, 60.57
bev  AP:93.96, 76.62, 69.59
3d   AP:89.41, 70.11, 63.32
aos  AP:86.56, 66.95, 60.12
So, has anyone reproduce the result claimed in the readme.md?

**Comment by 1453042287** [comment]
single gpu?

**Comment by zxduan-r** [comment]
yes, single gpu.

**Comment by skyhehe123** [comment]
Can you show me your log file

**Comment by zxduan-r** [comment]
This is my log. [screen.log](https://github.com/skyhehe123/SA-SSD/files/4398221/screen.log). Thanks.

**Comment by ZYang2077** [comment]
i want to ask the log file too :)

**Comment by skyhehe123** [comment]
Seems that your loss is not right, here is my log file
[20200322_182759.log](https://github.com/skyhehe123/SA-SSD/files/4398228/20200322_182759.log)

**Comment by zxduan-r** [comment]
OK, Thanks very much. I will follow the readme.md and run the scripts again. Also, have you tried to train on multiple gpus? Can you share the config of multiple gpus?

**Comment by skyhehe123** [comment]
You can simply run bash dist_train.sh in tools. Set --nproc_per_node to specify the #GPU

**Comment by ZYang2077** [comment]
@skyhehe123 So here is my question, how many GPU did you use when conducting your experiments? and the batch size per GPU. Thanks a lot!

**Comment by skyhehe123** [comment]
@AlphaPlusTT I only use one GPU with batch size 2

**Comment by ZYang2077** [comment]
that's impressive, great work！

**Comment by BB88Lee** [comment]
Do you have new result? Did you successfully reproduce the result?
@buaaJeremyduan

**Comment by zxduan-r** [comment]
@BB88Lee No, still trying. What about yours?

**Comment by BB88Lee** [comment]
Failed to reproduce the result. Got all None tensor after 2h of training. Haven't tried yet.
I am very curious if it is easy to reproduce the results successfully.
If you have new results, please tell me. @buaaJeremyduan

**Comment by ZYang2077** [comment]
@BB88Lee @buaaJeremyduan any progress？i use the default setting and single gpu got this：
Car AP@0.70, 0.70, 0.70:
bev  AP:96.69, 90.31, 87.65
3d   AP:93.20, 83.92, 81.00

**Comment by ZYang2077** [comment]
BTW, i use the spconv 1.0 not 1.1

**Comment by zwqnju** [comment]
I have also tried to train a model, and get the following evaluation result:

Car AP@0.70, 0.70, 0.70:
bbox AP:99.55, 93.79, 91.16
bev  AP:96.57, 90.41, 87.73
3d   AP:93.16, 84.06, 81.10
aos  AP:99.42, 93.42, 90.63

Car AP@0.70, 0.50, 0.50:
bbox AP:99.55, 93.79, 91.16
bev  AP:99.42, 95.97, 93.43
3d   AP:99.37, 95.86, 93.31
aos  AP:99.42, 93.42, 90.63

Similar with @AlphaPlusTT

**Comment by Son-Goku-gpu** [comment]
@skyhehe123 My similar results similar to others:

Car AP@0.70, 0.70, 0.70:
bbox AP:99.43, 93.77, 91.14
bev  AP:96.61, 90.37, 87.69
3d   AP:93.25, 84.27, 81.31
aos  AP:99.30, 93.47, 90.69

**Comment by skyhehe123** [comment]
@Son-Goku-gpu This result is fine, so I can close this issue.

**Comment by dllucas** [comment]
@buaaJeremyduan @BB88Lee Have you successfully reproduced the results?

**Comment by GuangshengShi** [comment]
@Son-Goku-gpu  the re-trained bev AP (90.37) is lower 2 points than published (92.79). 
@skyhehe123 I find nearly all re-trained results have worse results than published in readme.md. For  Any other reason?

**Comment by skyhehe123** [comment]
@Son-Goku-gpu Would you mind sharing your model?

**Comment by wdxpython** [comment]
> This is my log. [screen.log](https://github.com/skyhehe123/SA-SSD/files/4398221/screen.log). Thanks.

Sorry to disturb you. I ran into a bug while running the code. I added issues, where the train error contains my bug details. Have you encountered the same problem? If you have encountered it, how did you modify it?

**Comment by wdxpython** [comment]
> I have also tried to train a model, and get the following evaluation result:
> 
> Car AP@0.70, 0.70, 0.70:
> bbox AP:99.55, 93.79, 91.16
> bev AP:96.57, 90.41, 87.73
> 3d AP:93.16, 84.06, 81.10
> aos AP:99.42, 93.42, 90.63
> 
> Car AP@0.70, 0.50, 0.50:
> bbox AP:99.55, 93.79, 91.16
> bev AP:99.42, 95.97, 93.43
> 3d AP:99.37, 95.86, 93.31
> aos AP:99.42, 93.42, 90.63
> 
> Similar with @AlphaPlusTT

Sorry to disturb you. I ran into a bug while running the code. I added issues, where the train error contains my bug details. Have you encountered the same problem? If you have encountered it, how did you modify it?

**Comment by abhigoku10** [comment]
@AlphaPlusTT @GuangshengShi @Son-Goku-gpu @BB88Lee  hi team , what is the error your getting while training ? how long did it take to train ?i am not able to get the metrics as mentioned even after using the same split values, can you share the pkl files or is there any modifications to be done in test

**Comment by lihua213** [comment]
I use the 79th checkpoint_epoch to validate the model and get:
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3769/3769, 14.9 task/s, elapsed: 253s, ETA:     0sCar AP@0.70, 0.70, 0.70:
bbox AP:99.57, 96.07, 93.65
bev  AP:99.25, 93.24, 92.85
3d   AP:96.27, 89.19, 86.77
aos  AP:99.44, 95.83, 93.30
Car AP@0.70, 0.50, 0.50:
bbox AP:99.57, 96.07, 93.65
bev  AP:99.57, 96.16, 95.86
3d   AP:99.57, 96.13, 95.82
aos  AP:99.44, 95.83, 93.30

Very good method and thanks!
But in the training phase, there is a not fatal error,
in sigmoid_focal loss, pt = (1-pre_sigmoid)*target + pred_sigmoid*(1-target)
runtimeError: the size of tensor a (40954) must match the size of tensor b (40953) at no-singleton dimension 0.

**Comment by Divadi** [comment]
@lihua213 Hey, how did you resolve the error? my run crashed due to it

**Comment by lihua213** [comment]
@Divadi It is not a fatal error, so just run again.

**Comment by Ultra11nstinct** [comment]
@Divadi @lihua213 
sorry to bother, but have you solved the error?  i met the same error in epoch 60.

> I use the 79th checkpoint_epoch to validate the model and get:
> [>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3769/3769, 14.9 task/s, elapsed: 253s, ETA: 0sCar AP@0.70, 0.70, 0.70:
> bbox AP:99.57, 96.07, 93.65
> bev AP:99.25, 93.24, 92.85
> 3d AP:96.27, 89.19, 86.77
> aos AP:99.44, 95.83, 93.30
> Car AP@0.70, 0.50, 0.50:
> bbox AP:99.57, 96.07, 93.65
> bev AP:99.57, 96.16, 95.86
> 3d AP:99.57, 96.13, 95.82
> aos AP:99.44, 95.83, 93.30
> 
> Very good method and thanks!
> But in the training phase, there is a not fatal error,
> in sigmoid_focal loss, pt = (1-pre_sigmoid)_target + pred_sigmoid_(1-target)
> runtimeError: the size of tensor a (40954) must match the size of tensor b (40953) at no-singleton dimension 0.

**Comment by yangyongguang** [comment]
> @Divadi @lihua213
> sorry to bother, but have you solved the error? i met the same error in epoch 60.
> 
> > I use the 79th checkpoint_epoch to validate the model and get:
> > [>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3769/3769, 14.9 task/s, elapsed: 253s, ETA: 0sCar AP@0.70, 0.70, 0.70:
> > bbox AP:99.57, 96.07, 93.65
> > bev AP:99.25, 93.24, 92.85
> > 3d AP:96.27, 89.19, 86.77
> > aos AP:99.44, 95.83, 93.30
> > Car AP@0.70, 0.50, 0.50:
> > bbox AP:99.57, 96.07, 93.65
> > bev AP:99.57, 96.16, 95.86
> > 3d AP:99.57, 96.13, 95.82
> > aos AP:99.44, 95.83, 93.30
> > Very good method and thanks!
> > But in the training phase, there is a not fatal error,
> > in sigmoid_focal loss, pt = (1-pre_sigmoid)_target + pred_sigmoid_(1-target)
> > runtimeError: the size of tensor a (40954) must match the size of tensor b (40953) at no-singleton dimension 0.

have you solve this problem, i meet it every thrain, run it again do not help me.

**Comment by Ultra11nstinct** [comment]
> > @Divadi @lihua213
> > sorry to bother, but have you solved the error? i met the same error in epoch 60.
> > > I use the 79th checkpoint_epoch to validate the model and get:
> > > [>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3769/3769, 14.9 task/s, elapsed: 253s, ETA: 0sCar AP@0.70, 0.70, 0.70:
> > > bbox AP:99.57, 96.07, 93.65
> > > bev AP:99.25, 93.24, 92.85
> > > 3d AP:96.27, 89.19, 86.77
> > > aos AP:99.44, 95.83, 93.30
> > > Car AP@0.70, 0.50, 0.50:
> > > bbox AP:99.57, 96.07, 93.65
> > > bev AP:99.57, 96.16, 95.86
> > > 3d AP:99.57, 96.13, 95.82
> > > aos AP:99.44, 95.83, 93.30
> > > Very good method and thanks!
> > > But in the training phase, there is a not fatal error,
> > > in sigmoid_focal loss, pt = (1-pre_sigmoid)_target + pred_sigmoid_(1-target)
> > > runtimeError: the size of tensor a (40954) must match the size of tensor b (40953) at no-singleton dimension 0.
> 
> have you solve this problem, i meet it every thrain, run it again do not help me.

sorry, i didnt meet it anymore
sounds wierd ,but it exactly didnt crushed.

**Comment by Ultra11nstinct** [comment]
> @skyhehe123 My similar results similar to others:
> 
> Car AP@0.70, 0.70, 0.70:
> bbox AP:99.43, 93.77, 91.14
> bev AP:96.61, 90.37, 87.69
> 3d AP:93.25, 84.27, 81.31
> aos AP:99.30, 93.47, 90.69

@Son-Goku-gpu @AlphaPlusTT 
sorry to bother, your results seems similar to the published.
i wanna know do you have any modifications? can you share your config file?
and i wanna know that which checkpoint do you use, 50 or 80?
and do you use the pretrained model in training？

This is my results at checkpoint_50 with default config
Car AP@0.70, 0.70, 0.70:
bbox AP:99.19, 93.54, 91.00
bev  AP:96.71, 92.89, 90.39
3d   AP:91.80, 83.10, 78.70
aos  AP:99.14, 93.26, 90.41
Car AP@0.70, 0.50, 0.50:
bbox AP:99.19, 93.54, 91.00
bev  AP:99.30, 96.02, 93.56
3d   AP:99.29, 95.94, 93.46
aos  AP:99.14, 93.26, 90.41 

Looking forward to your reply

P.S. im a fan of Goku. 👍

**Comment by Son-Goku-gpu** [comment]
Hi, Goku fan @Ultra11nstinct  :)

I can only find this log (see the attached file) which I'm not sure if it corresponds my uploaded results above, but should be similar. In my memory, I did make several modifcations to make this codebase run and what I used is the previous version (50 epoches). As I just run it for several times and move to my own work, so I don't remember too much, and my suggestions for your trouble are as follows:
1. reproduce the info file, gt_database and other dataset files with this codebase, try not use the ones produced by SECOND codebase;
2. be careful for the spconv version (maybe 1.0, I forget but you can try)
3. be care of the AP_40 metric in this codebase, which uses 41 sampling points instead of 40 in the early version, making the published results a little larger (0.2 ~ 0.3 points);
4. the AP fluctuations of 0.5~0.8 is acceptable, better not try to seek the results strictly matched with the publication (only for your reference).

[log_0.txt](https://github.com/skyhehe123/SA-SSD/files/5344994/log_0.txt)

**Comment by Ultra11nstinct** [comment]
oh, i forgot to @Son-Goku-gpu

Thanks for your reply.

1. i have reproduced the db infos by using create_data.py for train and val sets.
2. i currently use spconv 1.2, you mean that maybe spconv 1.0 will get better performance? 
3. emm, i checked the evaluation code (kitti_eval.py) in master branch, and found it actually uses 41 sampling points. i didnt find 
    any code about "40 sample pts", is there any misconceive between us？
4. i agree with you. 
    My problem is that i got 3D_ap_moderate(83.10) by using the author's original work, 
    and i do several modifications to make the performance improved about 1 percent (84.07). 
    But it still lower than the author's Pretrained model for 84.54.
    i used the same codebase , same split file(same as SECOND,and i reproduce the db_infos for train and val) and same 
    epoch(50).
    really curious about why i cant reach the author's published work.

Thank you again for your carefully reply. :)
I will try spconv 1.0 to see if it has better performance.

**Comment by MuMuJun97** [comment]
@zxduan90 , sorry to bother

> I have also tried to train a model, and get the following evaluation result:
> 
> Car AP@0.70, 0.70, 0.70:
> bbox AP:99.55, 93.79, 91.16
> bev AP:96.57, 90.41, 87.73
> 3d AP:93.16, 84.06, 81.10
> aos AP:99.42, 93.42, 90.63
> 
> Car AP@0.70, 0.50, 0.50:
> bbox AP:99.55, 93.79, 91.16
> bev AP:99.42, 95.97, 93.43
> 3d AP:99.37, 95.86, 93.31
> aos AP:99.42, 93.42, 90.63
> 
> Similar with @AlphaPlusTT

When you retrained the network, did you use a **pre-trained model**?

https://github.com/skyhehe123/SA-SSD#pretrained-model 

In the Readme.txt, the author provide a pre-trained model. I don't know whether it should be use in the training progress. 

And what's the difference between the two results?  **AP@(11 recall points) vs 40 recall poisitions** ?
![image](https://user-images.githubusercontent.com/34757047/100298170-64be6d00-2fcb-11eb-8729-b237437dc61e.png)
![image](https://user-images.githubusercontent.com/34757047/100298200-730c8900-2fcb-11eb-8045-fa91e2c36e3f.png)

**Comment by vehxianfish** [comment]
@MuMuJun97 .
Hi,brother.Do you reproduce the result?If you did,please list your details.eg,data augmentaion,epochs,lib versions.
